File: Ruby on Rails/app/services/url_cache.rb
require "singleton"

class UrlCache
  include Singleton

  def initialize
    @mutex = Monitor.new
    @urls = {}
  end

  def load
    @mutex.synchronize do
      @urls = Url.pluck(:id, :original, :accesses).each_with_object({}) do |(id, original, accesses), hash|
        hash[id] = { "original" => original, "accesses" => accesses }
      end
    end
  end

  def get_all
    @mutex.synchronize { @urls.deep_dup }
  end

  def get(id)
    @mutex.synchronize { @urls[id]&.dup }
  end

  def exists?(id)
    @mutex.synchronize { @urls.key?(id) }
  end

  def set(id, info)
    @mutex.synchronize { @urls[id] = info }
  end

  def delete(id)
    @mutex.synchronize { @urls.delete(id) }
  end

  def increment_accesses(id)
    @mutex.synchronize do
      if @urls[id]
        @urls[id]["accesses"] += 1
        @urls[id]["accesses"]
      else
        0
      end
    end
  end
end

File: .NET/Encurtador.Url/Services/UrlCache.cs
using System.Collections.Concurrent;
using Encurtador.Url.Models;

namespace Encurtador.Url.Services;

public class UrlCache
{
    private readonly ConcurrentDictionary<string, UrlInfo> _urls = new();

    public void Load(IEnumerable<UrlItem> items)
    {
        _urls.Clear();
        foreach (var item in items)
        {
            _urls.TryAdd(item.Id, new UrlInfo(item.Original, item.Accesses));
        }
    }

    public Dictionary<string, UrlInfo> GetAll()
    {
        return _urls.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
    }

    public UrlInfo? Get(string id)
    {
        return _urls.TryGetValue(id, out var info) ? info : null;
    }

    public bool Exists(string id)
    {
        return _urls.ContainsKey(id);
    }

    public void Set(string id, UrlInfo info)
    {
        _urls[id] = info;
    }

    public void Delete(string id)
    {
        _urls.TryRemove(id, out _);
    }

    public long IncrementAccesses(string id)
    {
        var newInfo = _urls.AddOrUpdate(
            id,
            _ => new UrlInfo("", 1),
            (_, existing) => new UrlInfo(existing.Original, existing.Accesses + 1)
        );
        return newInfo.Accesses;
    }
}
